Avoid using mp.Manager to report #GPUs needed in dist tests

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61409**

We used a multiprocessing.Manager in order to share TEST_SKIPS between the parent and the child processes. TEST_SKIPS is a global variable that defines a unique error code for each "error type", so that the parent can figure out the reason a child exited. While originally this mapping was immutable, at some point we allowed children to modify the parent's value of that mapping so they could update the message for the `multi-gpu` error to make it reflect how many GPUs were really needed. This occurred in D23285790. Since then this Manager proved to be quite problematic, especially around thread safety, races, TSAN, ... (see D22753459, D23641618, D28490129, D28794321 and D29585862). This seems like an awful lot of trouble for such a small functionality. Here I propose we drop Manager and instead get the same result by using separate error codes for each number of GPUs. It should be much simpler and thus more robust.

Differential Revision: [D29612614](https://our.internmc.facebook.com/intern/diff/D29612614/)

lw committed:
Avoid using mp.Manager to report #GPUs needed in dist tests

We used a multiprocessing.Manager in order to share TEST_SKIPS between the parent and the child processes. TEST_SKIPS is a global variable that defines a unique error code for each "error type", so that the parent can figure out the reason a child exited. While originally this mapping was immutable, at some point we allowed children to modify the parent's value of that mapping so they could update the message for the `multi-gpu` error to make it reflect how many GPUs were really needed. This occurred in D23285790. Since then this Manager proved to be quite problematic, especially around thread safety, races, TSAN, ... (see D22753459, D23641618, D28490129, D28794321 and D29585862). This seems like an awful lot of trouble for such a small functionality. Here I propose we drop Manager and instead get the same result by using separate error codes for each number of GPUs. It should be much simpler and thus more robust.

Differential Revision: [D29612614](https://our.internmc.facebook.com/intern/diff/D29612614/)

[ghstack-poisoned]

pritamdamania87 commented:
Changes look good to me, although there are a few test failures. I guess we need a `multi-gpu-1` key too:

```
Jul 08 12:34:46 ERROR [0.016s]: test_allreduce_coalesced_checks_cuda (__main__.ProcessGroupGlooTest)
Jul 08 12:34:46 ----------------------------------------------------------------------
Jul 08 12:34:46 Traceback (most recent call last):
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 397, in wrapper
Jul 08 12:34:46     self._join_processes(fn)
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 596, in _join_processes
Jul 08 12:34:46     self._check_return_codes(elapsed_time)
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 636, in _check_return_codes
Jul 08 12:34:46     raise RuntimeError(error)
Jul 08 12:34:46 RuntimeError: Process 0 exited with error code 10 and exception:
Jul 08 12:34:46 Traceback (most recent call last):
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 515, in run_test
Jul 08 12:34:46     getattr(self, test_name)()
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 399, in wrapper
Jul 08 12:34:46     fn()
Jul 08 12:34:46   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 97, in wrapper
Jul 08 12:34:46     sys.exit(TEST_SKIPS[f'multi-gpu-{x}'].exit_code)
Jul 08 12:34:46 KeyError: 'multi-gpu-1'
```

lw committed:
Update on "Avoid using mp.Manager to report #GPUs needed in dist tests"


We used a multiprocessing.Manager in order to share TEST_SKIPS between the parent and the child processes. TEST_SKIPS is a global variable that defines a unique error code for each "error type", so that the parent can figure out the reason a child exited. While originally this mapping was immutable, at some point we allowed children to modify the parent's value of that mapping so they could update the message for the `multi-gpu` error to make it reflect how many GPUs were really needed. This occurred in D23285790. Since then this Manager proved to be quite problematic, especially around thread safety, races, TSAN, ... (see D22753459, D23641618, D28490129, D28794321 and D29585862). This seems like an awful lot of trouble for such a small functionality. Here I propose we drop Manager and instead get the same result by using separate error codes for each number of GPUs. It should be much simpler and thus more robust.

Differential Revision: [D29612614](https://our.internmc.facebook.com/intern/diff/D29612614/)

[ghstack-poisoned]
